### ISSUE TITLE
VM should treat const/final declarations as not introducing a setter.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/sdk#28267]
Example:
```dart
const x = 42;  // same with "final".
set x(v) { print(v); }
main() {
  x = "Hello world!";
}
```
The language semantics specifies that this should call the setter and print the string. Instead the VM reports:
```
'file:///test/concon.dart': error: line 2 pos 5: Variable 'x' is already defined in this library
set x(v) { print(v); }
```
Final declarations, and const declarations are final, don't introduce a setter. The setter (line 2) introduces a setter named `x=` which is not related to final/const variable.

**Comment by rmacnak-google** [comment]
Circa 2013, immutable fields were specified as introducing a setter that threw. Has this changed?

**Comment by lrhn** [comment]
Yes, that was dropped again in https://github.com/dart-lang/sdk/commit/7ba09adde9ade9e733b600c0bd318c9e088a1d53 (Oct 2013).

**Comment by natebosch** [comment]
I think this Dart 2 changed this and the error is intended.

**Comment by eernstg** [comment]
I just checked (2.1.0-dev.2.0) `dart` and `dartanalyzer`, and they both reject the [example](https://github.com/dart-lang/sdk/issues/28267#issue-198975614) with several error messages.

However, even though Dart 2 changes the rules about conflicts involving setters and other entities, the situation which is described here is not an error (and wasn't ever since 7ba09ad), so we should keep this issue alive, and make sure that the tools do not emit any errors.

I'm reopening this issue and creating two sub-issues, one for the vm and another one for the analyzer (such that each issue has just one associated task). I'll also edit the initial text of this issue to add the check boxes for each subissue.

**Comment by natebosch** [comment]
I think in this case we should change the spec rather than the tools.

I can't think of a valid reason to allow creating a setter with the same name as a const variable. If we do allow this we'd probably need a lint to add back the error that exists now and enforce it in google internal code...

**Comment by matanlurey** [comment]
Strongly agree re: change the spec.

**Comment by leafpetersen** [comment]
Is there a reason that we want this to be allowed?

**Comment by lrhn** [comment]
A const variable is necessarily static (or top-level), so it only affects the class itself. I don't see a particular reason to want a setter, it's just never been prohibited by the specification.

So, yes, we can definitely make const variable declarations special, and treat them specially in look-ups, differently from non-const variables, but I'd actually have thought that to be more work, not less.
If that work has already been done (consistently), then changing the spec is fine with me.

Maybe it's enough to just say that a scope cannot have two declarations with the same base name unless one is a setter and the other is a getter which does not come from a constant variable declaration.

**Comment by natebosch** [comment]
> For _final_ fields, I think we do need to allow a setter.

For instance fields the tools currently allow it. This is valid:

```dart
class Foo {
  final x = 1;
  set x(int x) => print(x);
}
```

For top level or statics the definition is allowed but usage is not - I think the definition shouldn't be allowed either.

**Comment by lrhn** [comment]
I see no reason to prevent the usage for statics (top-level or class), and I definitely don't want to make a difference between instance fields and static fields wrt. something like this. Consistency is too valuable to drop it without a very good reason.
Since the members are static, they can be resolved statically, so there shouldn't be any issue recognizing this case.

**Comment by eernstg** [comment]
I agree with @lrhn that we shouldn't introduce a special exception for these particular cases (e.g., it would certainly be an inconsistency to allow an instance setter to coexist with a final instance variable, but disallow a static or top-level setter to coexist with a corresponding final variable).

But it seems even more un-Dart-2-ish to accept the negation of the title: 'VM should treat const/final declarations as introducing a setter', because this implies that assignments to a const/final variable would be a _dynamic_ error rather than a static error. Do you really want that?!

**Comment by natebosch** [comment]
I hope for both the definition and usage of a setter with the same name as a const variable to be a static error, however we phrase the issue to get that outcome is fine with me.

As for the inconsistency I don't think most developers will be confused when presented with a difference between static and instance variables. This issue is the first time I ever saw it come up and our tool has disallowed the usage for a long time. Google code will almost certainly disallow this even if the language allows it.

**Comment by eernstg** [comment]
So you are basically saying that we should make it a compile-time error to _declare_ a setter along with a final or const library-or-static variable? (But it's still OK with an instance variable?)

**Comment by natebosch** [comment]
Yes that's what I would prefer. I feel most strongly about const, I think static final should probably also disallow it too.

**Comment by lrhn** [comment]
Can you say *why* you want to disallow a setter declared alongside a final static variable?

It shouldn't be an implementation issue - it's all static anyway.

I can see why instance variables can be seen as different, because you can also inherit a setter from a super-class or add it in a sub-class. The declarations don't need need to be in the same class to give the same effect, and we likely won't prevent the inherited case, so it makes litte sense to prevent the adjacent case.
Even so, I'd probably prefer to either disallow setters and final fields being declared in the same scope everywhere or nowhere, not make a difference between instance and static scopes. That way, you can either declare a field, or you can declare getter and/or setter, but never a field and a setter.
(That would make it easier to later change the getter/setter syntax in the future, say to something like C#'s `int foo {get => _foo; set(v) { _foo = v; }}`, which would be an alternative to a field declaration).

**Comment by natebosch** [comment]
> I'd probably prefer to either disallow setters and final fields being declared in the same scope everywhere or nowhere

I think it'd probably be fine to disallow it everywhere other than the fact that it's a breaking change. I'm also receptive to arguments around allowing it due to the reasons you stated with inheritance.

We have a nice opportunity for disallowing it for statics since that's the current tool behavior so it isn't breaking to keep doing that. We shouldn't backslide to allowing it everywhere since that makes it even harder to disallow in Dart 3.

> Can you say _why_ you want to disallow a setter declared alongside a final static variable?

We've been disallowing it and no one has complained. Other than consistency with instance variables, why do we want to allow it? What code does it make more readable and understandable?

It's not a pattern I want to see in code. If it's allowed at the language level we'll need a separate lint to prevent it.

**Comment by leafpetersen** [comment]
> I think it'd probably be fine to disallow it everywhere other than the fact that it's a breaking change.

+1

Allowing final fields + setters just seems like a bad idea for anything, but not fixable for instance fields now.

**Comment by eernstg** [comment]
Based on the principle that orthogonal language features (that is, "all combinations are supported") are better than a hodgepodge of exceptions, I'd prefer to have simple and consistent rules, even when some special cases seem like a bad idea, and then we're perfectly free to flag such special cases with the linter.

Besides, who says that the following couldn't be considered meaningful?:

```dart
// ----------------------------------------------------------------------

// Model a mutable variable.
final varM = ProtectedValue(0);
set varM(ProtectedValue v) => varM.update(v);

// Model a final variable.
final varF = ProtectedValue(100);

main() {
  var client = ClientOfProtectedValue();

  // We can use a protected value as a variable, final or mutable.
  print(client.foo(varM));
  varM = varF;
  // varF = ...; // Error
  print(client.foo(varM));
}

// ----------------------------------------------------------------------
// Different library: _x is not accessible for outside clients.

class ProtectedValue<X> {
  X _x;
  ProtectedValue(this._x);
  void update(ProtectedValue<X> newValue) {
    // ... hard work, validating the new value for `_x` ...
    _x = newValue._x;
  }
}

class ClientOfProtectedValue {
  foo(ProtectedValue<int> pInt) {
    return 2 * pInt._x; // Think 'one-way hash function', not leaking _x.
  }
}
```
Of course, it's still possible to use `varF.update(...)` to "assign to `varF`", but I'm sure we can come up with a variant that uses an extra `ReadOnlyProtectedValue` class or something like that to enforce immutability (and that one would of course not have an `update` method, so it's statically enforced).

I'm not saying that you'd love to do this, just that we might not be well advised to cut it off at the language level just because we haven't seen all the ways that it might be used.

**Comment by natebosch** [comment]
> better than a hodgepodge of exceptions, I'd prefer to have simple and consistent rules

The difference between static final and static instance variables is already somethings users have to keep in mind.

- Will this be initialized on first read or not?
- Can this be overridden?
- Can this be declared as `const`?

> I'm sure we can come up with a variant that uses an extra `ReadOnlyProtectedValue` class or something like that to enforce immutability

I think the pattern we'd use is the same one I'd recommend to work around being unable to add a setter with the same name as a static final variable.

```dart
final _varM = ProtectedValue(0);
ReadOnlyProtectedValue get varM => _varM;
```

Same as I'd do

```dart
const _x = 42;
int get x => _x;
set x(int v) { print(v); }
```

The value is that weird code looks weird, instead of slipping by in code review.


This thread has discussed consistency - are there other principles that would lead us to a different decision? How about the principle of least surprise? Everyone I've talked to who looked at the code in the first comment was surprised that it's something we'd expect to work.

**Comment by eernstg** [comment]
I adjusted the code [here](https://github.com/dart-lang/sdk/issues/28267#issuecomment-418736790) slightly, in order to model final variables (providing a `Value`) and mutable variables (offering r/w access known from a `Variable`) directly, which turned out to be straightforward. So there's `ProtectedValue` and `ProtectedVariable` now, and the _latter_ is the one that used be named `ProtectedValue`.

@natebosch wrote:

> The difference between static final and static instance variables is already somethings users
> have to keep in mind.

Indeed, the semantics differ in crucial ways, and developers must know about it. I'm just saying that we shouldn't introduce gratuitous differences.

I think the following snippet combines the ideas from the two snippets you show:

```dart
final _varM = ProtectedValue(0);
ReadOnlyProtectedValue get varM => _varM;
set varM(ProtectedValue v) { _varM.update(v); }
```

This shows that you can work around the error on the combination final var + setter by inserting a getter between the two. Agreed, but this just shows that we aren't getting protection from anything, we're just requiring developers who might want to do such a thing to write some boilerplate.

**Comment by leafpetersen** [comment]
We're not going to change the spec on this.  Fixing the bug is low priority, but we should try to ensure that the fix (when it happens) is done consistently across implementations.

**Comment by eernstg** [comment]
Closing, given that the spec was not updated in response to this issue. The specified behavior is also the implemented behavior for almost all configurations (exception: [DDC](https://github.com/dart-lang/sdk/issues/63477)).
